Box value types implementing `IXmlSerializable` in `XmlSerializer` generated IL

daeghanelkin opened the pull request:
Value types implementing the `IXmlSerializable` interface were not being property boxed when they declared a parameterless constructor. This caused the `XmlSerializer` to fail to deserialize them with an `InvalidProgramException` exception when the generated IL tried to call the `XmlSerializationReader.ReadSerializable` method with the value.

Fixes #99613

---

I would like to note that I am not really that familiar with IL generation, so all corrections are welcome ☺️.

I took the C# output of the serializer produced by the `Microsoft.XmlSerializer.Generator` tool for a different types and simplified it down to determine what the IL should look like for the call to the `XmlSerializationReader.ReadSerializable` method.

My simplified snippet was the following:

```csharp
public class Class {
    public void Method() {
        Read(new StructWithParameterlessConstructor());

        Read(new StructWithParameterlessConstructor(), true);
        
        Read(new StructWithoutParameterlessConstructor());

        Read(new StructWithoutParameterlessConstructor(), true);
    }
    
    private static void Read(IInterface @interface) { }
    
    private static void Read(IInterface @interface, bool b) { }
}

public interface IInterface;

public struct StructWithParameterlessConstructor : IInterface
{
    public StructWithParameterlessConstructor() { }
}

public struct StructWithoutParameterlessConstructor : IInterface;
```

The following is the resulting IL for each method:

<table>
<tr>
<th>C#</th>
<th>IL</th>
</tr>
<tr>
<td>

```csharp
Read(new StructWithParameterlessConstructor())
```

</td>
<td>

```csharp
newobj instance void StructWithParameterlessConstructor::.ctor()
box StructWithParameterlessConstructor
call void Class::Read(class IInterface)
```

</td>
</tr>
<tr>
<td>

```csharp
Read(new StructWithParameterlessConstructor(), true)
```

</td>
<td>

```csharp
newobj instance void StructWithParameterlessConstructor::.ctor()
box StructWithParameterlessConstructor
ldc.i4.1
call void Class::Read(class IInterface)
```

</td>
</tr><tr>
<td>

```csharp
Read(new StructWithoutParameterlessConstructor())
```
</td>
<td>

```csharp
ldloca.s 0
initobj StructWithoutParameterlessConstructor
ldloc.0
box StructWithoutParameterlessConstructor
call void Class::Read(class IInterface)
```

</td>
</tr>
<tr>
<td>

```csharp
Read(new StructWithoutParameterlessConstructor(), true)
```

</td>
<td>

```csharp
ldloca.s 0
initobj StructWithoutParameterlessConstructor
ldloc.0
box StructWithoutParameterlessConstructor
ldc.i4.1
call void Class::Read(class IInterface)
```

</td>
</tr>
</table>

No boxing was being written by the IL generator for the value types with parameterless constructors when the XML serializer was generating the runtime assembly, so a simple addition of that boxing seems to resolve the issue.
Value types without parameterless constructors do not run into this issue because they fall into the `sm.TypeDesc.CannotNew` condition above the new condition I've added, and avoid the style of instantiation that's demonstrated in my simplified example.

filipnavara commented:
[area-Serialization](https://github.com/dotnet/runtime/labels/area-Serialization) recently changed ownership, but it can be more appropriate.  cc @StephenMolloy

daeghanelkin committed:
Box value types implementing `IXmlSerializable` in `XmlSerializer` generated IL

Value types implementing the `IXmlSerializable` interface were not being
property boxed when they declared a parameterless constructor. This
caused the `XmlSerializer` to fail to deserialize them with an
`InvalidProgramException` exception when the generated IL tried to call
the `XmlSerializationReader.ReadSerializable` method with the value.

Fixes [#99613](https://github.com/dotnet/runtime/issues/99613)

StephenMolloy reviewed (approved):
LGTM.

Just for clarity, this is a scenario which did not work in .Net 4.8. But I think it's one that should. So here we go.

Also another fun fact - this already works with the reflection-based serializer in .Net Core.

(Finally, it's taken me quite a while to get back to serialization. So if the `Expected.SerializableAssembly.XmlSerializers.cs` is out of date, I'll create a separate PR to fix that up.)
